File: src/main/java/net/arvandor/numinoustreasury/command/numinousitem/NuminousItemCommand.java
package net.arvandor.numinoustreasury.command.numinousitem;

import com.rpkit.core.service.Services;
import net.arvandor.numinoustreasury.item.NuminousItemService;
import net.arvandor.numinoustreasury.item.NuminousItemStack;
import net.arvandor.numinoustreasury.item.NuminousItemType;
import net.arvandor.numinoustreasury.item.log.NuminousLogEntry;
import net.md_5.bungee.api.chat.BaseComponent;
import net.md_5.bungee.api.chat.TextComponent;
import net.md_5.bungee.chat.ComponentSerializer;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

import java.time.Instant;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

import static net.md_5.bungee.api.ChatColor.*;

public final class NuminousItemCommand implements CommandExecutor, TabCompleter {

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (!(sender instanceof Player player)) {
            sender.sendMessage(RED + "You must be a player to create items.");
            return true;
        }
        if (!sender.hasPermission("numinoustreasury.command.numinousitem")) {
            sender.sendMessage(RED + "You do not have permission to create items.");
            return true;
        }
        if (args.length < 1) {
            sender.sendMessage(RED + "Usage: /" + label + " [item type] (amount)");
        }
        NuminousItemService itemService = Services.INSTANCE.get(NuminousItemService.class);
        String itemName = String.join(" ", args);
        int amount = 1;
        NuminousItemType itemType = itemService.getItemTypeById(itemName);
        if (itemType == null) {
            itemType = itemService.getItemTypeByName(itemName);
        }
        if (itemType == null && args.length > 1) {
            itemName = String.join(" ", Arrays.stream(args).limit(args.length - 1).toArray(String[]::new));
            itemType = itemService.getItemTypeById(itemName);
            if (itemType == null) {
                itemType = itemService.getItemTypeByName(itemName);
            }
            try {
                amount = Integer.parseInt(args[args.length - 1]);
            } catch (NumberFormatException exception) {
                itemType = null;
            }
        }
        if (itemType == null) {
            sender.sendMessage(RED + "There is no item by that name.");
            return true;
        }
        NuminousItemStack numinousItem = new NuminousItemStack(
                itemType,
                amount,
                List.of(
                        new NuminousLogEntry(
                                Instant.now(),
                                player.getUniqueId(),
                                true,
                                new BaseComponent[] {
                                        new TextComponent("Created via command")
                                }
                        )
                )
        );
        ItemStack bukkitItem = numinousItem.toItemStack();
        player.getInventory().addItem(bukkitItem).values().forEach(overflowItem -> player.getWorld().dropItem(player.getLocation(), overflowItem));
        sender.sendMessage(GREEN + "Created " + amount + " × " + itemType.getName());
        return true;
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String label, String[] args) {
        NuminousItemService itemService = Services.INSTANCE.get(NuminousItemService.class);
        if (args.length == 0) {
            return itemService.getItemTypes().stream().flatMap(item -> Stream.of(item.getId(), item.getName())).toList();
        } else {
            return itemService.getItemTypes().stream().flatMap(item -> Stream.of(item.getId(), item.getName())).filter(name -> name.toLowerCase().startsWith(String.join(" ", args).toLowerCase())).toList();
        }
    }
}

File: src/main/kotlin/net/arvandor/numinoustreasury/command/numinousitem/NuminousItemCommand.kt
package net.arvandor.numinoustreasury.command.numinousitem

import com.rpkit.core.service.Services
import net.arvandor.numinoustreasury.NuminousTreasury
import net.arvandor.numinoustreasury.item.NuminousItemService
import net.arvandor.numinoustreasury.item.NuminousItemStack
import net.arvandor.numinoustreasury.item.log.NuminousLogEntry
import net.arvandor.numinoustreasury.mixpanel.NuminousMixpanelService
import net.arvandor.numinoustreasury.mixpanel.event.NuminousMixpanelItemCreatedEvent
import net.md_5.bungee.api.ChatColor
import net.md_5.bungee.api.chat.BaseComponent
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.entity.Player
import java.time.Instant

class NuminousItemCommand(private val plugin: NuminousTreasury) : CommandExecutor, TabCompleter {
    override fun onCommand(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): Boolean {
        if (sender !is Player) {
            sender.sendMessage(ChatColor.RED.toString() + "You must be a player to create items.")
            return true
        }
        if (!sender.hasPermission("numinoustreasury.command.numinousitem")) {
            sender.sendMessage(ChatColor.RED.toString() + "You do not have permission to create items.")
            return true
        }
        if (args.size < 1) {
            sender.sendMessage(ChatColor.RED.toString() + "Usage: /" + label + " [item type] (amount)")
        }
        val itemService =
            Services.INSTANCE.get(
                NuminousItemService::class.java,
            )
        var itemName = args.joinToString(" ")
        var amount = 1
        var itemType = itemService.getItemTypeById(itemName)
        if (itemType == null) {
            itemType = itemService.getItemTypeByName(itemName)
        }
        if (itemType == null && args.size > 1) {
            itemName = args.dropLast(1).joinToString(" ")
            itemType = itemService.getItemTypeById(itemName)
            if (itemType == null) {
                itemType = itemService.getItemTypeByName(itemName)
            }
            try {
                amount = args.last().toInt()
            } catch (exception: NumberFormatException) {
                itemType = null
            }
        }
        if (itemType == null) {
            sender.sendMessage(ChatColor.RED.toString() + "There is no item by that name.")
            return true
        }
        val numinousItem =
            NuminousItemStack(
                itemType,
                amount,
                null,
                listOf(
                    NuminousLogEntry(
                        Instant.now(),
                        sender.uniqueId,
                        true,
                        arrayOf<BaseComponent>(
                            TextComponent("Created via command"),
                        ),
                    ),
                ),
            )
        val bukkitItem = numinousItem.toItemStack()
        sender.inventory.addItem(bukkitItem).values.forEach { overflowItem ->
            sender.world.dropItem(
                sender.location,
                overflowItem,
            )
        }
        sender.sendMessage(ChatColor.GREEN.toString() + "Created " + amount + " × " + itemType.name)

        val mixpanelService =
            Services.INSTANCE.get(
                NuminousMixpanelService::class.java,
            )
        val finalItemType = itemType
        val finalAmount = amount
        plugin.server.scheduler.runTaskAsynchronously(
            plugin,
            Runnable {
                mixpanelService.trackEvent(NuminousMixpanelItemCreatedEvent(sender, finalItemType, finalAmount, "Command"))
            },
        )
        return true
    }

    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): List<String> {
        val itemService =
            Services.INSTANCE.get(
                NuminousItemService::class.java,
            )
        return if (args.isEmpty()) {
            itemService.itemTypes.flatMap { item ->
                listOf(
                    item.id,
                    item.name,
                )
            }
        } else {
            itemService.itemTypes.flatMap { item ->
                listOf(
                    item.id,
                    item.name,
                )
            }.filter { name: String ->
                name.lowercase()
                    .startsWith(args.joinToString(" ").lowercase())
            }
        }
    }
}
